### PR TITLE
fix(moq-gst): keep one rendition's subscribe off the catalog loop

### PR DESCRIPTION
**Pull request by kixelated**
## Summary

- **Root cause**: `reconcile` awaited `broadcast.track(name)?.subscribe(None)` serially *inside* the catalog-follow loop. That await resolves only once the track info arrives, so a rendition nobody serves parked the whole loop: no other rendition started and no later catalog update was ever read.
- The browser publisher announces an audio-only catalog a beat before its video encoder configures, so `moqsrc` subscribed `audio`, parked, and never saw the update that added `video`. That is the intermittent `FAIL js -> gst` in the smoke matrix (run [33124111425](https://github.com/moq-dev/moq/actions/runs/33124111425)): the cell burned its full 30s while every other subscriber finished within 4s, and the relay log shows that session subscribing `catalog.json` then `audio` and never `video`.
- `reconcile` is now synchronous and infallible; each pump awaits its own subscription and creates its pad only once the track is live, so a pad downstream can link to is a promise the rendition is flowing.

Everything below came out of review, all the same defect class: one rendition that never arrives must not degrade the ones that do.

- **A rendition the broadcast refuses** propagated out of `reconcile` and ended the session. It now skips just that rendition, matching how an unsupported codec or a malformed init were already handled.
- **Pad ids were claimed when a pump was spawned**, so an unserved video rendition reserved `video_0` and parked while the rendition that arrived became `video_1` — breaking the documented `moqsrc name=s s.video_0 ! ...` link, including the smoke pipeline this PR moves onto it. The id is claimed at pad creation instead.
- **Teardown and pad creation could both win.** A `PumpState` compare-exchange settles `SUBSCRIBING -> LIVE` against `SUBSCRIBING -> CANCELLED`, so a cancel landing mid-subscription can't leave a pad exposed and then yanked without an EOS.
- `run_pump` reached eight arguments, so its inputs moved into a `Pump` struct with a `run` method.

**Harness** (`test/smoke/smoke.sh`), two fixes: `moqsrc ! filesink` linked whichever Sometimes pad appeared first, so an audio-carrying publisher could pass the gst cell on **audio** bytes with video never flowing; it now links `s.video_0` by name, the form `doc/bin/gstreamer.md` and `demo/sub/justfile` already prescribe. Each cell also reports its elapsed seconds, which is what separates a stall from a slow start — the failing run would have read `FAIL js -> gst (30s of 30s)` beside `PASS js -> rust (2s)`.

Deliberately **not** done: no timeout bump and no cell retry. The budget was never the constraint, and a retry would have hidden a bug that also affects anyone pointing `moqsrc` at a browser publisher.

## Public API changes

None. Everything is private to `rs/moq-gst`: `reconcile` (now sync, infallible), the new `follow_catalog` extracted from `run_session`, and `run_pump` becoming `Pump::run`. No pad names, properties, or element behaviour visible to a pipeline changed.

## Test plan

Five hermetic regression tests in `source::imp::session_tests`, driving `follow_catalog` against an in-process `broadcast::Producer` with no relay:

- `a_rendition_nobody_serves_does_not_block_the_others` — waits for the pending subscription through the broadcast's `Dynamic` before announcing video, so it can't pass by reading a combined snapshot in a lucky `plan.add` order.
- `a_refused_rendition_does_not_end_the_session` — both renditions in one snapshot, so neither outcome depends on ordering.
- `an_unserved_rendition_does_not_claim_the_first_pad_name` — same-kind: an unanswered video, then a live one, asserting the live pad takes the first name.
- `a_subscription_resolving_after_cancellation_creates_no_pad`.
- `a_closing_catalog_keeps_the_renditions_it_named` — guards the reverted attempt below.

Re-created the previous behaviour and ran the suite 5x: the four that target it fail on every run.

Also `just check`, `just test` (120 passed), `just test smoke-full` (21/21, gst cells 0–2s), and `just test smoke-negative`. A forced failure reports `FAIL js -> gst (2s of 2s)`.

Known gap: the `PumpState` race needs `cancel` and `subscribe` ready in the same poll, which can't be forced without test hooks in production code. Covered indirectly by the late-resolution test.

## Attempted and reverted

An earlier commit cancelled not-yet-live pumps when the catalog closed, to stop an unanswered rendition holding the session open. That is wrong: "hasn't taken a pad yet" is indistinguishable from "hasn't been polled yet", and a publisher that names its tracks and then finishes the catalog delivers the snapshot and the close back to back — so it dropped the entire broadcast. Reverted in 2a67f0731, with the reproduction kept as a regression test. The termination problem it was aimed at is #3139; it is not a regression from this PR, since the same input previously parked `reconcile` inside the loop.

## Cross-package sync

`rs/moq-gst` -> `doc/bin/gstreamer.md` not updated: no flag, property, or pad-naming change, and the docs already teach the `moqsrc name=s s.video_0 ! ...` form this PR moves the harness onto.

## Follow-ups

- #3137 — bound how many pending rendition subscriptions a catalog can open.
- #3139 — a rendition nobody answers keeps the session alive after the catalog closes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

(Written by Claude Opus 5)

### CODE REVIEW

File: rs/moq-gst/src/source/imp.rs [hunk 1]
```diff
@@ -526,16 +549,35 @@ impl TrackDescriptor {
...
 	mut cancel: watch::Receiver<bool>,
 ) {
+	// Resolves once the publisher answers with the track info. A catalog can name a track its
+	// publisher never serves, so this can wait forever; racing `cancel` keeps such a pump
+	// reapable, and holding the wait here rather than in `reconcile` keeps it off every other
+	// rendition.
+	let subscriber = tokio::select! {
+		_ = cancel.changed() => return,
+		subscriber = track.subscribe(None) => match subscriber {
+			Ok(subscriber) => subscriber,
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-gst/src/source/imp.rs]: Agreed and fixed in ea3553942: the per-kind id is now claimed in `TrackDescriptor::claim` at pad creation, after the subscription resolves, so a rendition that never arrives reserves nothing. Added a same-kind regression (unanswered video, then a live video) asserting the live pad takes the first name; it reproduces the reported `video_0`/`video_1` split without the fix.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: rs/moq-gst/src/source/imp.rs [hunk 2]
```diff
@@ -800,3 +842,132 @@ mod tests {
...
+			guard.audio.renditions = BTreeMap::from([("audio".to_string(), audio_rendition())]);
+		}
+
+		let (shutdown, mut shutdown_rx) = watch::channel(false);
+		let consumer = broadcast.consume();
+		let weak = element.downgrade();
+		let session = super::RUNTIME.spawn(async move { follow_catalog(consumer, weak, &mut shutdown_rx).await });
+
+		// Second update adds video, backed by a real track so its subscription resolves.
+		let _video = broadcast.create_track("video", None).unwrap();
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-gst/src/source/imp.rs:1051]: Agreed and fixed in daba066d6. The test now waits for the pending audio subscription through the broadcast's `Dynamic` and holds the request unanswered before announcing video, so the session provably reconciled the audio-only snapshot rather than possibly skipping to a combined one and winning on `plan.add` ordering. The sibling refused-rendition test announces both renditions in a single snapshot for the same reason. Re-ran both against the previous behaviour 5x: 4/4 tests fail on every run, where the stalled-rendition one was previously order-dependent.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: rs/moq-gst/src/source/imp.rs [hunk 3]
```diff
@@ -466,7 +484,14 @@ async fn reconcile(
 		};
 		let (cancel_tx, cancel_rx) = watch::channel(false);
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-gst/src/source/imp.rs:531]: Not changing this one. The `watch::channel(false)` cancel is pre-existing, not new here: it is on `origin/main` at `rs/moq-gst/src/source/imp.rs:467`, alongside the same pattern for `SessionController`'s shutdown. This PR moves where it is awaited, not what it is. `moq-gst` has no `kio` dependency at all and is described in rs/CLAUDE.md as bridging to a background tokio task, which it owns a runtime for, so swapping both channels to kio is a crate-wide refactor rather than part of a bug fix. Happy to see it done separately if the convention should extend here.

Separately, the DoS concern from the same review is filed as #3137 with the reasoning; the catalog-close termination point you flagged is fixed in ea3553942 and daba066d6.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: rs/moq-gst/src/source/imp.rs [hunk 4]
```diff
@@ -353,26 +371,32 @@ async fn run_session(
...
+					//
+					// A pump still awaiting its subscription is the exception. No further
+					// catalog update can arrive to release it and the publisher has stopped
+					// describing the broadcast, so it would park forever and hold the session
+					// open with it. It has no pad, so there is no EOS for downstream to miss.
+					None => {
+						catalog_closed = true;
+						for track in active.values() {
+							if !track.live.load(Ordering::Relaxed) {
+								let _ = track.cancel.send(true);
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-gst/src/source/imp.rs]: Agreed, and this was worse than described. Reproduced it in a test: a publisher that names its tracks and then finishes the catalog has the snapshot and the close arrive back to back, so the close was routinely observed before the freshly spawned pumps ran at all, and every rendition it had just announced was cancelled before it could subscribe. That dropped the whole broadcast, not just a race on one pump.

Reverted the catalog-close cancellation in 2a67f0731 and kept the reproduction as `a_closing_catalog_keeps_the_renditions_it_named`. Tracking subscription state separately would not have fixed it: "not resolved yet" is equally indistinguishable from "not polled yet". The underlying termination problem is filed as #3139 with the options, since separating the two needs a signal the catalog does not carry. Not a regression either way -- before this branch the same input parked `reconcile` inside the loop.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: rs/moq-gst/src/source/imp.rs [hunk 5]
```diff
@@ -294,6 +330,25 @@ struct ActiveTrack {
...
+	state: Arc<PumpState>,
+}
+
+impl ActiveTrack {
+	/// Tear the pump down whatever stage it reached: one still subscribing never takes a pad,
+	/// one that has drops it when it sees the watch.
+	fn cancel(&self) {
+		self.state.cancel_before_live();
+		let _ = self.cancel.send(true);
+	}
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-gst/src/source/imp.rs]: Agreed, fixed in c26808e51. Both callers already own the `ActiveTrack` (one from `active.drain()`, one from `active.remove()`), so taking `self` costs nothing and makes a second teardown unrepresentable.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: rs/moq-gst/src/source/imp.rs [hunk 6]
```diff
@@ -526,51 +621,101 @@ impl TrackDescriptor {
...
+			}
+		};
+		let mut track = moq_mux::container::Consumer::new(subscriber, container).with_latency(Duration::from_secs(1));
+
+		// Winning this is what earns a pad. Losing means a teardown got here while the
+		// subscription was still resolving (this rendition was removed, reshaped, or outlived by
+		// a closing catalog), and it must not publish a pad at all: the watch alone can't say
+		// that, since a cancel landing just after we read it would leave a pad exposed and then
+		// yanked without an EOS.
+		if !state.go_live() {
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-gst/src/source/imp.rs:666]: Not changing this one. `go_live()` is the commit point: winning it is what makes the pump the owner of a pad, and cancellation after that goes through the live-pump path on purpose. The end state you describe (pad removed without an EOS) is what cancelling a live pump already does at any point in its life, including one that streamed for an hour before a reshape removed it, so this race changes how briefly the pad existed rather than introducing a new outcome. Closing the window properly means holding a lock across `obj.add_pad()`, i.e. across a GStreamer call that can block and re-enter, which is a real hazard traded for a cosmetic gain.

Whether cancellation should emit EOS before dropping a pad is a fair question, but it applies to every cancellation path and predates this PR, so it belongs in its own change rather than bolted onto this race.

Worth noting the near-identical suggestion one commit earlier is what caused the media loss in #3139: a partial fix to a race here, justified by reasoning rather than a failing test, is exactly the move that went wrong. I would rather leave a documented commit point than narrow this window without a test that can demonstrate the difference.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: rs/moq-gst/src/source/imp.rs [hunk 7]
```diff
@@ -800,3 +945,277 @@ mod tests {
...
+	}
+
+	/// Block until a consumer asks the broadcast for a track, returning the request unanswered so
+	/// its subscriber stays parked. Bounded so a session that never subscribes fails the test.
+	fn await_request(dynamic: &mut moq_net::broadcast::Dynamic) -> moq_net::track::Request {
+		super::RUNTIME
+			.block_on(async { tokio::time::timeout(Duration::from_secs(10), dynamic.requested_track()).await })
+			.expect("no track was ever requested")
+			.expect("broadcast closed")
+	}
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-gst/src/source/imp.rs:1003]: Not changing this one. `tokio::time::pause()` panics on a multi-threaded runtime, and this element's `RUNTIME` is `new_multi_thread` because the pump needs `block_in_place` for `pad.push()`, so the suggested remedy would fail outright here rather than help.

The guideline targets tests whose behaviour depends on timers firing. These do not: the 10s in `await_request` and the bounded poll in `await_pad` are failure bounds that never elapse on a passing run, so CI load changes how long a *failure* takes to report, not whether the test passes. `await_pad` is a plain `std::thread::sleep` poll loop, not a tokio timer at all.

The 500 ms sleep is the one real wait, and it is asserting a negative. Unlike `await_pad` there is no edge to synchronize on: it gives the runtime a window in which the un-cancelled version reliably creates a pad. Added a comment saying so in c26808e51.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: rs/moq-gst/src/source/imp.rs [hunk 8]
```diff
@@ -526,51 +609,101 @@ impl TrackDescriptor {
...
+		let mut track = moq_mux::container::Consumer::new(subscriber, container).with_latency(Duration::from_secs(1));
+
+		// Winning this is what earns a pad. Losing means a teardown got here while the
+		// subscription was still resolving (this rendition was removed, reshaped, or outlived by
+		// a closing catalog), and it must not publish a pad at all: the watch alone can't say
+		// that, since a cancel landing just after we read it would leave a pad exposed and then
+		// yanked without an EOS.
+		if !state.go_live() {
+			return;
+		}
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-gst/src/source/imp.rs:668]: Not changing this one, same reasoning as the identical finding CodeRabbit raised on this line and then withdrew (r3882479009 -> r3882499620).

`go_live()` is the ownership commit point. Winning it is what makes the pump the pad's owner, and cancellation after that deliberately takes the live-pump path. The outcome you describe, a pad removed without EOS, is what cancelling a live pump already does at any point in its life: a rendition that streamed for an hour before a reshape removed it ends exactly the same way. So this ordering changes how briefly the pad existed, not what downstream observes. There is no distinct failure mode to regress against, which is also why the suggested test would assert current intended behaviour rather than a bug.

Making publication atomic with cancellation means holding a lock across `obj.add_pad()`, a GStreamer call that can block and re-enter, which is a genuine hazard traded for a cosmetic gain.

Whether cancellation should emit EOS before dropping a pad is a real question, but it applies to every live-pump cancellation path and predates this PR, so it wants its own change.

Worth flagging why I am being conservative here specifically: the previous commit tightened a race in this same area on reasoning rather than a failing test, and that is what caused the media loss now tracked in #3139. I would rather leave a documented commit point than narrow this window the same way.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
